### ISSUE TITLE
missing index on temporary table in "repair"

### ISSUE THREAD

**Issue by kunom** [duplicati/duplicati#1569]
Recreating a local database seems to be massively CPU bound. I have a backup repository of approx. 25'000 files and 10 revisions, and `restore` runs for minutes.

When looking with Process Explorer at the call stack of the main thread, it sits most of the time at

```
11	System.Data.SQLite.dll!System.Data.SQLite.SQLiteCommand.ExecuteScalar+0x16
12	Duplicati.Library.Main.dll!Duplicati.Library.Main.Database.LocalRecreateDatabase.UpdateBlockset+0x6e
```

The query of interest is:
```
m_findblocklisthashCommand.CommandText = string.Format(@"SELECT DISTINCT ""BlockListHash"" FROM ""{0}"" WHERE ""BlockListHash"" = ? ", m_tempblocklist);
m_findblocklisthashCommand.AddParameters(1);
```

Not only looks this to me like an EXISTS check, it also seems that adding an index to the following temporary table (at LocalRecreateDatabase.cs:90) could drastically improve performance:

```
            using (var cmd = m_connection.CreateCommand())
                cmd.ExecuteNonQuery(string.Format(@"CREATE TEMPORARY TABLE ""{0}"" (""BlockListHash"" TEXT NOT NULL, ""BlockHash"" TEXT NOT NULL, ""Index"" INTEGER NOT NULL)", m_tempblocklist));
```

**Comment by kenkendk** [comment]
I don't see how an EXISTS check could make it faster?

Pull requests are always welcome :)

**Comment by kunom** [comment]
No, not faster, but maybe clearer. However, this issue is about the missing index. I'll give it a try.

**Comment by kenkendk** [comment]
I have tested this fix, and it works, but I have not measured performance:
https://github.com/duplicati/duplicati/tree/suggested_fix_for_issue_1569

**Comment by kunom** [comment]
I will check and give you feedback. Thanks.

**Comment by kunom** [comment]
Please be aware that you missed to provide the name of the temporary table in the new statement, which leads to a FormatException at startup.

I did some measurements. Recreating the database seems to be divided into three major phases, summing up in my case to 1060 seconds:
- Downloading and processing few large files (163 seconds)
- Downloading and processing many small files (810 seconds)
- Final processing without screen output (85 seconds)

With the new index on the temporary table, the second phase shrinks to 66 seconds, the total execution time down to 315 seconds. And, in addition, the second phase seems not any more entirely CPU bound.

**Comment by kenkendk** [comment]
Arh, that explains why the unittests are failing :).

But that is a great speedup. I will fix the commit and merge it in.

**Comment by kenkendk** [comment]
This was fixed in #1580.
